example/pub-sub-chat.ss

```diff
@@ -0,0 +1,143 @@
+        :std/getopt
+        :std/sugar
+        :std/iter
+        :std/logger
+        :std/net/bio
+        :vyzo/libp2p
+        :std/text/base64
+        :std/text/utf8
+        :std/misc/channel)
+(export main)
+
+
+;;Main
+;;After parsing the commandline arguments, this will call do-startup or display an error for incorrect arguments
+(def (main . args)
+  (def bootstrap-cmd
+    (command 'bootstrap help: "listen for incoming connections"
+               (argument 'host-addresses help: "Comma separated multi addresses")))
+  (def connect-cmd
+    (command 'connect help: "connect to a bootsrap"
+             (argument 'peer value: string->peer-info help: "peer to initially connect to")
+             (argument 'host-addresses  help: "Comma separated multi addresses")))
+  (def help-cmd
+    (command 'help help: "display help; help <command> for command help"
+             (optional-argument 'command value: string->symbol)))
+  (def gopt
+    (getopt bootstrap-cmd connect-cmd help-cmd))
+
+  (try
+   (let ((values cmd opt) (getopt-parse gopt args))
+     (let-hash opt
+       (case cmd
+         ((bootstrap) ;;In case where the command is to bootstrap
+          (do-startup #f .host-addresses))
+         ((connect) ;;where command is to connect
+          (do-startup .peer .host-addresses))
+         ((help) ;;where command is to help
+          (getopt-display-help-topic gopt .?command "pub-sub-chat")))))
+   (catch (getopt-error? exn)
+     (getopt-display-help exn "pub-sub-chat" (current-error-port))
+     (exit 1))))
+
+
+;; do-startup
+;; Parameters:
+;; peer: #f or a peer Addresss
+;; host-addresses: a list of multiaddresses
+;;
+;;If peer is #f, do-startup will not try to connect to any peers and will subscribe to the topic "chat"
+;;
+;;If peer is anything else, do-startup will try to connect to peer and will subscribe to the topic "chat"
+(def (do-startup peer host-addresses)
+   (displayln "starting up")
+
+   ;start the daemon with custom arguments -pubsub and -connManager to enable pubsub
+   (start-libp2p-daemon! host-addresses: host-addresses options: ["-pubsub" "-connManager"] wait: 10)
+
+   ;open a client with the given host-address, it will use the daemon already running
+  (let* ((c (open-libp2p-client host-addresses: host-addresses wait: 10))
+         (self (libp2p-identify c)))
+    (for (p (peer-info->string* self))
+      (displayln "I am " p)) ;Identify yourself. Print each multiaddr
+
+    (if peer ;If there is a bootstrap peer (connecting command) then connect to it
+      (begin
+        (displayln "Bootsrap peer found, connecting to it")
+        (libp2p-connect c peer)
+        (displayln "Connected to Bootstrap peer")))
+
+    ;Subscribe to the chat and spawn a reader thread and writer thread
+    (let*-values (((sub cancel) (pubsub-subscribe c "chat"))
+           (reader (spawn subscription-reader sub self)))
+      (subscription-writer c)
+      (thread-terminate! reader)))) ;;TODO: Does this exit cleanly?
```

Comment:
no, it really abruptly kills the thread -- you might want to raise an exception in it to interrupt it. See `:std/misc/threads`, you can use `thread-raise!` or `thread-abort!`.

Reply:
Sounds good! Is there a way to make sure the daemon exits cleanly as well? Or will this do it?

Reply:
Just pushed your suggestions!

Reply:
i think it should be fine, but might make sense to explicitly close the client somewhere.